Add a progress bar based on `ProgressMeter.jl`

DilumAluthge opened the pull request:
Fixes #135

This pull request adds the `ProgressMeterReport`, which uses the `ProgressMeter.jl` package to display a progress bar.

To use this reporter, simply set the `reporter` keyword argument when calling `mcmc_with_warmup`:
```julia
reporter = ProgressMeterReport()
```

cc: @tpapp

DilumAluthge commented:
Alright, this is ready for review now @tpapp

tpapp commented:
Thanks. I am very busy now but will try to do it soon. Geel free to bump this if I don't get to it within a week.

tpapp commented on src/reporting.jl:
```diff
@@ -34,7 +36,7 @@ Return a reporter which can be used for progress reports with a known number of
 key-value pairs.
 """
-make_mcmc_reporter(reporter::NoProgressReport, total_steps; meta...) = reporter
+make_mcmc_reporter(reporter::NoProgressReport, total_steps; currently_warmup::Bool=false, meta...) = reporter
```
```suggestion
make_mcmc_reporter(reporter::NoProgressReport, total_steps; currently_warmup::Bool = false, meta...) = reporter
```
(spaces)

tpapp commented on src/reporting.jl:
```diff
@@ -98,7 +100,7 @@ function report(reporter::LogMCMCReport, message::AbstractString; meta...)
 end
 
-function make_mcmc_reporter(reporter::LogProgressReport, total_steps::Integer; meta...)
+function make_mcmc_reporter(reporter::LogProgressReport, total_steps::Integer; currently_warmup::Bool=false, meta...)
```
```suggestion
function make_mcmc_reporter(reporter::LogProgressReport, total_steps::Integer; currently_warmup::Bool = false, meta...)
```

tpapp commented on src/reporting.jl:
```diff
@@ -34,7 +36,7 @@ Return a reporter which can be used for progress reports with a known number of
 `total_steps`. May return the same reporter, or a related object. Will display `meta` as
```
please document currently_warmup in the docstring

DilumAluthge replied:
Done!

tpapp reviewed (changes requested):
LGTM, I just requested some trivial changes. Thanks for adding this nice feature!

DilumAluthge commented:
I think I've addressed all the changes!

You will also notice that I have now removed the `log_warmup` field. This bears some explanation.

Right now, we have separate progress bars for warmup and MCMC. Long-term, I think the best solution will be to figure out how to have a single progress bar for both warmup and MCMC. This will require some more info, e.g. https://github.com/tpapp/DynamicHMC.jl/issues/141

Once we implement that feature, the `log_warmup` field will no longer make sense. If we add the `log_warmup` field now, then later, when we are ready to remove it, that would be a breaking change.

So I think it's best to right now remove the `log_warmup` field, and always generate progress bars for both warmup and MCMC. And then in the future we can work on combining the progress bars for warmup and MCMC into a single progress bar.

tpapp commented:
Thanks, I think this is the best solution for now.

I am merging this, but if that is OK with you I will wait for JuliaNLSolvers/Optim.jl#863 to resolve before tagging a new release. If you need that before, just let me know.

DilumAluthge commented:
> I am merging this, but if that is OK with you I will wait for [JuliaNLSolvers/Optim.jl#863](https://github.com/JuliaNLSolvers/Optim.jl/issues/863) to resolve before tagging a new release. If you need that before, just let me know.

Sounds good to me, there's no time pressure on my part.
